zerver/decorator.py

```diff
@@ -481,7 +486,15 @@ def _wrapped_view_func(
         request: HttpRequest, /, *args: ParamT.args, **kwargs: ParamT.kwargs
     ) -> HttpResponse:
         process_client(request, request.user, is_browser_view=True, query=view_func.__name__)
-        rate_limit(request)
+
+        if not should_rate_limit(request):
+            return view_func(request, *args, **kwargs)
+
+        if request.user.is_authenticated:
+            rate_limit_user(request, request.user, domain="api_by_user")
+        else:
+            rate_limit_request_by_ip(request, domain="api_by_ip")
```

Comment:
This concluding API feels unclean; why not juave have `should_rate_limit` be called inside `rate_limit_user` and `rate_limit_request_by_ip`? It'd be really nice to get away from having multiple "return view_func(...)" lines in these decorators.

Reply:
We can do that. The only issue here will be that this check is normally not done for other callers of `rate_limit_request_by_ip`. To maintain the old behavior, we can probably add a flag that allows the caller to skip the check.

Reply:
I think we can probably allow these checks as an improvement. It will ensure that the internal clients do get exempt from rate limiting.

Reply:
This is implemented in c6b08b1d737ca01d76be4ef69df5ffd131c2ca96.

Reply:
Yeah, @mateuszmandera FYI, but it's probably a bug that we weren't doing these checks for `rate_limit_request_by_ip`.